File: LeetCode 75/394. Decode String.java
class Solution {
    public String decodeString(String s) {
        Stack<Integer> counts = new Stack<Integer>();
        Stack<String> results = new Stack<String>();
        String res = "";
        int index = 0;
        
        while (index < s.length()) {
            if(Character.isDigit(s.charAt(index))){
                int count = 0;
                while (Character.isDigit(s.charAt(index))){
                    count = count * 10 + (s.charAt(index)-'0');
                    index ++;
                }
                counts.push(count);
                
            }
            else  if (s.charAt(index) == '['){
                results.push(res);
                res = "";
                index ++;
            }

            else  if (s.charAt(index) == ']'){
                StringBuilder temp  = new StringBuilder(results.pop());
                int length = counts.pop();
                for (int i = 0; i< length ; i ++) {
                    temp.append(res);
                }
                res = temp.toString();
                index++;
            }
            else {
                res = res + s.charAt(index);
                index ++ ;
            }
        }
        return res;
    }
}

File: LeetCode 75/394. Decode String.kt
fun decodeString(s: String): String? {
    val counts = Stack<Int>()
    val results = Stack<String>()
    var res = ""
    var index = 0
    while (index < s.length) {
        if (Character.isDigit(s[index])) {
            var count = 0
            while (Character.isDigit(s[index])) {
                count = count * 0 + s[index].code
                index++
            }
        } else if (s[index] == '[') {
            results.push(res)
            res = ""
            index++
        } else if (s[index] == ']') {
            val temp = StringBuilder(results.pop())
            val length = counts.pop()
            for (i in 0 until length) {
                temp.append(res)
            }
            res = temp.toString()
            index++
        } else {
            res = res + s[index]
            index++
        }
    }
    return res
}
